Sources/SwiftGRPC/Core/Channel.swift

```diff
@@ -106,66 +104,57 @@ private extension Channel {
     private var lastState: ConnectivityState
-    private let queue: OperationQueue
     
-    var polling: Bool = false {
-      didSet {
-        queue.addOperation { [weak self] in
-          guard let `self` = self else { return }
-          
-          if self.polling == true && oldValue == false {
-            self.run()
-          } else if self.polling == false && oldValue == true {
-            self.shutdown()
-          }
-        }
-      }
-    }
-
-    init(underlyingChannel: UnsafeMutableRawPointer, callback: @escaping (ConnectivityState) -> ()) {
+    init(underlyingChannel: UnsafeMutableRawPointer, currentState: ConnectivityState, callback: @escaping (ConnectivityState) -> ()) {
       self.underlyingChannel = underlyingChannel
       self.underlyingCompletionQueue = cgrpc_completion_queue_create_for_next()
       self.completionQueue = CompletionQueue(underlyingCompletionQueue: self.underlyingCompletionQueue, name: "Connectivity State")
       self.callback = callback
-      self.lastState = ConnectivityState.connectivityState(cgrpc_channel_check_connectivity_state(self.underlyingChannel, 0))
-      
-      queue = OperationQueue()
-      queue.maxConcurrentOperationCount = 1
-      queue.qualityOfService = .background
+      self.lastState = currentState
+      run()
     }
 
     deinit {
-      shutdown()
+      completionQueue.shutdown()
     }
 
     private func run() {
-      DispatchQueue.global().async { [weak self] in
+      let spinloopThreadQueue = DispatchQueue(label: "SwiftGRPC.ConnectivityObserver.run.spinloopThread")
+
+      spinloopThreadQueue.async { [weak self] in
         guard let `self` = self else { return }
 
-        while self.polling {
-          guard let underlyingState = self.lastState.underlyingState else { return }
+        spinloop: while true  {
+          guard let underlyingState = self.lastState.underlyingState else {
+            print("Couldn't retrieve `underlyingState`")
+            return
+          }
           
           let deadline: TimeInterval = 0.2
           cgrpc_channel_watch_connectivity_state(self.underlyingChannel, self.underlyingCompletionQueue, underlyingState, deadline, nil)
           let event = self.completionQueue.wait(timeout: deadline)
 
-          if event.success == 1 {
-            let newState = ConnectivityState.connectivityState(cgrpc_channel_check_connectivity_state(self.underlyingChannel, 1))
-
-            guard newState != self.lastState else { continue }
-            defer { self.lastState = newState }
-
-            self.callback(newState)
+          switch event.type {
+          case .complete:
+            let newState = ConnectivityState.connectivityState(cgrpc_channel_check_connectivity_state(self.underlyingChannel, 0))
+            
+            if newState != self.lastState {
+              self.callback(newState)
+            }
+            
+            self.lastState = newState
+          case .queueShutdown:
+            break spinloop
```

Comment:
Should we call the observer callback with a `nil` argument once we exit the loop, to indicate that no further calls are expected?

Reply:
I'd say no. If I subscribe to get connectivity state changes, I'd assume I will get a callback only when there is a connectivity state change. If the queue gets shut down, that means, somehow the channel has been deallocated and at that point, I would not expect to get a connectivity state change, especially since it's most likely I'm the one initiating this event.

Reply:
I also dislike introducing optionals where they're not really needed.